### PR TITLE
.Net: Add TextEmbedding based on Microsoft.ML

### PR DESCRIPTION
**Pull request by weldpua2008**
### Motivation and Context

Add Microsoft.ML TextEmbedding based on Microsoft.ML.Tokenizers library for offline use.

### Description

Add offline TextEmbedding based on Microsoft.ML.Tokenizers

### Contribution Checklist
 * [x]  The code builds clean without any errors or warnings
 * [x]  The PR follows the [SK Contribution Guidelines](https://github.com/microsoft/semantic-kernel/blob/main/CONTRIBUTING.md) and the [pre-submission formatting script](https://github.com/microsoft/semantic-kernel/blob/main/CONTRIBUTING.md#development-scripts) raises no violations
 * [x]  All unit tests pass, and I have added new tests where possible
 * [x]  I didn't break anyone 😄

[Relevant Issue](https://github.com/microsoft/semantic-kernel/issues/478)

### CODE REVIEW

**Comment by lemillermicrosoft** [comment]
Cool idea! Tagging #2840 for additional consideration.

File: dotnet/src/Connectors/Connectors.MicrosoftML/TextEmbedding/MicrosoftMLTextEmbeddingGeneration.cs [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+    /// <param name="loggerFactory">Optional logger factory for logging.</param>
+    public MicrosoftMLTextEmbeddingGeneration(
+        string vocabFilePath,
+        string mergeFilePath,
+        ILoggerFactory? loggerFactory = null
+    )
+    {
+        this.vocabFilePath = vocabFilePath;
+        this.mergeFilePath = mergeFilePath;
+        this.tokenizer = new Tokenizer(new Bpe(vocabFilePath, mergeFilePath));
```
Comment (lemillermicrosoft) [inline dotnet/src/Connectors/Connectors.MicrosoftML/TextEmbedding/MicrosoftMLTextEmbeddingGeneration.cs:47]: I wonder if it'd be more flexible to just supply a Tokenizer object so consumers could supply a RobertaPreTokenizer, etc.

File: dotnet/src/Connectors/Connectors.MicrosoftML/MicrosoftMLTextKernelBuilderExtensions.cs [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+{
+    /// <summary>
+    /// Configures the MsMLTextTextEmbeddingGenerationService for text embedding generation.
+    /// </summary>
+    /// <param name="builder">The <see cref="KernelBuilder"/> instance</param>
+    /// <param name="serviceId">Optional service ID.</param>
+    /// <param name="vocabFilePath">The JSON file path containing the dictionary of string keys and their ids.</param>
+    /// <param name="mergeFilePath">The file path containing the tokens's pairs list.</param>
+    /// <param name="setAsDefault">Whether the service should be the default for its type.</param>
+    public static KernelBuilder WithMicrosoftMLTextTextEmbeddingGenerationService(this KernelBuilder builder,
```
Comment (lemillermicrosoft) [inline dotnet/src/Connectors/Connectors.MicrosoftML/MicrosoftMLTextKernelBuilderExtensions.cs:23]:
```suggestion
    public static KernelBuilder WithMicrosoftMLTextEmbeddingGenerationService(this KernelBuilder builder,
```

File: dotnet/src/Connectors/Connectors.MicrosoftML/TextEmbedding/MicrosoftMLTextEmbeddingGeneration.cs [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+    /// <summary>
+    /// Generates embeddings from tokenizer tokens.
+    /// </summary>
+    /// <param name="token">The tokenizer result.</param>
+    /// <param name="cancellationToken">Optional cancellation token.</param>
+    /// <returns>A ReadOnlyMemory of float representing the embeddings.</returns>
+    private static ReadOnlyMemory<float> GenerateEmbeddingsFromTokens(TokenizerResult token, CancellationToken cancellationToken)
+    {
+        // Convert token.Ids (int) to float
+        var floatIds = token.Ids.Select(id => (float)id).ToArray();
```
Comment (lemillermicrosoft) [inline dotnet/src/Connectors/Connectors.MicrosoftML/TextEmbedding/MicrosoftMLTextEmbeddingGeneration.cs:80]: These embeddings need to be normalized to allow for various embeddings calculations.

File: dotnet/src/Connectors/Connectors.MicrosoftML/TextEmbedding/MicrosoftMLTextEmbeddingGeneration.cs (hunk 1 above)
Comment (lemillermicrosoft) [reply dotnet/src/Connectors/Connectors.MicrosoftML/TextEmbedding/MicrosoftMLTextEmbeddingGeneration.cs:47]: I went ahead and took a stab at adding this. Highlights that there is issue with embeddings calculations.

File: dotnet/samples/KernelSyntaxExamples/Example14_SemanticMemory.cs [hunk 4]
```diff
@@ -64,6 +89,61 @@ public static async Task RunAsync()
...
+
+        /* You can also use a local embedding generator.
+         *
+         * In this example we use a local embedding generator that uses a local
+         * embedding generator based on Microsoft.ML tokenizers.
+         */
+
+        var kernelWithLocalEmbeddings = Kernel.Builder
+            .WithLoggerFactory(ConsoleLogger.LoggerFactory)
+            .WithMicrosoftMLTextEmbeddingGenerationService(serviceId: "roberta", setAsDefault: true) // TODO Files are missing
```
Comment (lemillermicrosoft) [inline dotnet/samples/KernelSyntaxExamples/Example14_SemanticMemory.cs:106]: TODO (also change serviceId when it's done)

File: dotnet/samples/KernelSyntaxExamples/Example14_SemanticMemory.cs [hunk 5]
```diff
@@ -64,6 +89,61 @@ public static async Task RunAsync()
...
+            throw new System.IO.FileNotFoundException("Missing required resources");
+        }
+
+        Microsoft.ML.Tokenizers.EnglishRoberta model = new(encoder, vocab, dict);
+
+        Microsoft.ML.Tokenizers.Tokenizer tokenizer = new(model, new Microsoft.ML.Tokenizers.RobertaPreTokenizer());
+
+        var kernelWithLocalEmbeddings = Kernel.Builder
+            .WithLoggerFactory(ConsoleLogger.LoggerFactory)
+            .WithMicrosoftMLTextEmbeddingGenerationService(tokenizer: tokenizer, serviceId: "roberta", setAsDefault: true) // TODO Array lengths must be equal
```
Comment (lemillermicrosoft) [inline dotnet/samples/KernelSyntaxExamples/Example14_SemanticMemory.cs:140]: TODO -- I believe something needs to be done to normalize the embeddings for cosine similarity comparison. Right now, these embeddings cannot be used. Maybe BPE is okay but could you please add some example files to use for that to the sample project with examples @weldpua2008 ? Thank you!

File: dotnet/src/Connectors/Connectors.MicrosoftML/TextEmbedding/MicrosoftMLTextEmbeddingGeneration.cs (hunk 3 above)
Comment (lemillermicrosoft) [reply dotnet/src/Connectors/Connectors.MicrosoftML/TextEmbedding/MicrosoftMLTextEmbeddingGeneration.cs]: Tracking in new conversation.

File: dotnet/src/Connectors/Connectors.AI.MicrosoftML/TextEmbedding/MicrosoftMLTextEmbeddingGeneration.cs [hunk 6]
```diff
@@ -0,0 +1,93 @@
...
+    /// <param name="loggerFactory">Optional logger factory for logging.</param>
+    public MicrosoftMLTextEmbeddingGeneration(
+        string vocabFilePath,
+        string mergeFilePath,
+        ILoggerFactory? loggerFactory = null
+    )
+    {
+        this._tokenizer = new Tokenizer(new Bpe(vocabFilePath, mergeFilePath));
+        this.Logger = loggerFactory is not null ? loggerFactory.CreateLogger(this.GetType()) : NullLogger.Instance;
+    }
```
Comment (WeihanLi) [inline dotnet/src/Connectors/Connectors.AI.MicrosoftML/TextEmbedding/MicrosoftMLTextEmbeddingGeneration.cs:41]:
```suggestion
    : this (new Tokenizer(new Bpe(vocabFilePath, mergeFilePath)), loggerFactory)
    {
    }
```

**Comment by markwallace-microsoft** [comment]
@weldpua2008 We are working on an update memory connector strategy for Semantic Kernel. @dmytrostruk will publish an ADR with the new approach soon. Let's review this PR to see how it aligns with this strategy when the ADR is published.

**Comment by markwallace-microsoft** [comment]
Closing this and we will revisit as part of the updated memory connector plan

**Pull request closed without merging**